### PR TITLE
Add fuzzer.

### PR DESCRIPTION
**Pull request by finnbear**
I use this excellent library in 3 projects and wanted to ensure it won't panic :)

## Changes
- [x] Add fuzzer
  - [x] Build a BVH
  - [x] Assert consistent
  - [x] Assert tight.
  - [x] Flatten the BVH
  - [x] Traverse the BVH
    - [x] Normal
    - [x] Iterator
    - [x] Nearest
    - [x] Farthest
    - [x] Flat
    - [ ] ~~Assert all traversals output the same set of shapes~~ doesn't work, possibly due to rounding errors
    - [x] Assert nearest and farthest traversals output the same set of shapes
    - [x] Assert normal and iterator traversals output the same results when the ray doesn't barely hit/miss any AABB
  - [x] Mutate a BVH
    - [x] Add shape(s)
    - [ ] ~~Remove shape(s)~~ causes internal panics
- [x] Run fuzzer in CI, for 5 minutes
- [x] `cargo check`/`cargo fmt` fuzzer in CI
- [x] Document fuzzer
- [ ] ~~Don't panic on empty BVH~~ (merge #116 insted)
  - [ ] ~~`assert_consistent` returns successfuly on an empty BVH~~ (merge #116 insted)
  - [ ] ~~`assert_tight` returns successfully on an empty BVH~~ (merge #116 insted)
  - [ ] ~~`is_consistent` returns `true` on an empty BVH~~ (merge #116 insted)
  - [ ] ~~`flatten` returns an empty flat BVH on an empty BVH~~ (merge #116 insted)
  - [ ] ~~Nearest/farthest iterators work on empty BVH~~ (merge #112 instead)

## Future work
- Fix internal panic in `remove_shape`.
- Make nearest/farthest/flat traversals output the same results as normal/iterator traversals when the ray doesn't barely hit/miss any AABB

### CODE REVIEW

**Comment by svenstaro** [comment]
Very cool! Before giving this a more in-depth review: We're already using proptest which has a somewhat related goal (trying a bunch of constrained values). I feel we should perhaps only have one of these.

Also, I really like that you fixed the empty BVH case. However, I think that should be a separate PR (and a bunch of tests to show that it's fixed).

What do you think?

**Comment by finnbear** [comment]
> We're already using proptest which has a somewhat related goal (trying a bunch of constrained values). I feel we should perhaps only have one of these.

There is some overlap here, but fuzzing has a distinct advantage because it recombines inputs to maximize code coverage, which can find edge-cases more efficiently than trying random inputs. IMO it's fine to have both, with `proptest` being used for writing manual tests of specific properties and fuzzing being used to automatically find gaps in those manual tests. Once the fuzzer finds something (like empty BVH panics), it should graduate to a normal unit test or `proptest`.


> Also, I really like that you fixed the empty BVH case. However, I think that should be a separate PR (and a bunch of tests to show that it's fixed).

Sure, I can move that change to another PR. It will make the fuzzer fail until the fixes are merged into `main` and this PR.

**Comment by svenstaro** [comment]
Thanks! I merged all the things. Feel free to rebase.

**Review by svenstaro — CHANGES_REQUESTED** [review]

File: fuzz/fuzz_targets/fuzz.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let _traverse_iterator = bvh
+                .traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let _traverse_flat = flat_bvh.traverse(&ray, &self.shapes);
+
+            // Fails, either due to bug or rounding errors.
+            // assert_eq!(traverse, traverse_iterator);
```
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs:152]: After this PR is merged, I will look into these. However, I don't think a fix is likely to exist, because each iterator/traversal uses slightly different ray-aabb intersections which will have different rounding errors for rays that just barely hit/miss a leaf AABB or node AABB.

I think a similar test could succeed on a modified BVH in which all AABB's are in a 3D grid, with their centers on integer coordinates, if each ray starts on an integer coordinate and travels in the X, Y, or Z direction.

File: fuzz/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "bvh-fuzz"
+version = "0.0.0"
+authors = ["Automatically generated"]
+publish = false
+edition = "2018"
+
+[package.metadata]
```
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/Cargo.toml:8]: Done documenting. Can consider `Justfile` in separate PR.

File: fuzz/fuzz_targets/fuzz.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{Hash, Hasher};
+
+use arbitrary::Arbitrary;
+use bvh::aabb::{Aabb, Bounded};
+use bvh::bounding_hierarchy::{BHShape, BoundingHierarchy};
+use bvh::bvh::Bvh;
+use bvh::ray::Ray;
+use libfuzzer_sys::fuzz_target;
+use nalgebra::{Point, SimdPartialOrd};
```
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs]: Good idea; done!

File: fuzz/fuzz_targets/fuzz.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs]: Update: I just added that "similar test" and it actually makes this line pass. However, all other commented-out traversals can still disagree, which strongly implies the existence of bug(s).

File: fuzz/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (svenstaro) [reply fuzz/Cargo.toml:8]: That works for me!

File: fuzz/fuzz_targets/fuzz.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+struct Workload<const D: usize> {
+    shapes: Vec<ArbitraryShape<D>>,
+    ray: ArbitraryRay<D>,
+    mutations: Vec<ArbitraryMutation<D>>,
+}
+
+impl<const D: usize> Workload<D> {
+    /// Called directly from the `cargo fuzz` entry point. Code in this function is
+    /// easier for `rust-analyzer`` than code in that macro.
+    fn fuzz(mut self) {
```
Comment (svenstaro) [inline fuzz/fuzz_targets/fuzz.rs:210]: Perhaps make the reader aware to look at the module level documentation to get the gist of what we're trying to do in this function.

File: fuzz/fuzz_targets/fuzz.rs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+            > 32
+        {
+            // Prevent traversal stack overflow by limiting max BVH depth to the traversal
+            // stack size limit.
+            return;
+        }
+
+        loop {
+            let assert_traversal_agreement =
+                self.ray.mode.is_grid() && self.shapes.iter().all(|s| s.mode.is_grid());
```
Comment (svenstaro) [inline fuzz/fuzz_targets/fuzz.rs:229]: What's the idea behind this?

File: fuzz/fuzz_targets/fuzz.rs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+                .into_iter()
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+
+            if assert_traversal_agreement {
+                assert_eq!(traverse, traverse_iterator);
+                // Fails, due to a bug.
+                // assert_eq!(traverse, traverse_flat);
+            } else {
+                // Fails, probably due to rounding errors.
```
Comment (svenstaro) [inline fuzz/fuzz_targets/fuzz.rs:256]: Let's add TODOs for the failures. Perhaps even add an additional tracking issue to look into these?

File: fuzz/fuzz_targets/fuzz.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            let farthest_traverse_iterator = bvh
+                .farthest_traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+
+            if assert_traversal_agreement {
+                // Fails, due to a bug.
+                // assert_eq!(traverse_iterator, nearest_traverse_iterator);
+            } else {
+                // Fails, probably due to rounding errors.
```
Comment (svenstaro) [inline fuzz/fuzz_targets/fuzz.rs:272]: Like above, some TODOs to track these failures would be great.

File: fuzz/fuzz_targets/fuzz.rs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    }
+}
+
+impl<T> Eq for ByPtr<'_, T> {}
+
+impl<T> Hash for ByPtr<'_, T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_usize(self.0 as *const _ as usize);
+    }
+}
```
Comment (svenstaro) [inline fuzz/fuzz_targets/fuzz.rs:320]: This feels a bit like a crutch and it seems there should be a better way for us to compare these but I don't really have a better idea here either.

**Review by svenstaro — CHANGES_REQUESTED** [review]
I finished the review. I think this is very cool and I've never used structural fuzzing this way as I'm usually content with just proptesting. I do see the point in this case, however.

I'm fine with the implementation but noted some parts where I'd like _even more_ comments. ;)

File: fuzz/fuzz_targets/fuzz.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+
+impl<const D: usize> ArbitraryRay<D> {
+    /// Produces the corresponding ray from the input.
+    fn ray(&self) -> Ray<Float, D> {
+        // Double normalize helps when the first one encounters precision issues.
+        let mut direction = (self.destination.point() - self.origin.point())
+            .normalize()
+            .normalize();
```
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs:133]: `normalize` here is an `nalgebra` function with a very standard implementation of (approximately) `self / sqrt(self.x^2 + self.y^2 + ...)`. Precision issues are to be expected when used with extremely large inputs that `Arbitrary` is happy to produce. The first normalize produces error depending on the input size. The second normalize has almost no error.

File: fuzz/fuzz_targets/fuzz.rs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+        // Ensure no degenerate direction.
+        if direction.magnitude() < 0.5 || direction.iter().any(|f| f.is_nan() || f.abs() > 1.5) {
+            direction.iter_mut().for_each(|f| *f = 1.0);
+            direction = direction.normalize();
+        }
+        assert!(
+            direction.magnitude() - 1.0 < 0.1,
+            "{}",
+            direction.magnitude()
+        );
```
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs:143]: First of all, this fuzzing code could be written better as much of it is redundant. I will do that now.

Passing a vector with NaN, Infinity, or -Infinity components (what this was *supposed* to test for) is invalid input. If you want to protect users, we could make `Ray::new` panic if the direction has non-finite components.

File: fuzz/fuzz_targets/fuzz.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs]: This code actually normalized 3 times, 2 before `Ray::new` and 1 in `Ray::new`. In the current state of the code, we don't need to pre-normalize at all; the `Ray::new` normalization is sufficient. I think I wrote this kludge before limiting the arbitrary coordinate magnitudes to something reasonable.

File: fuzz/fuzz_targets/fuzz.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs]: This code became unnecessary after I started limiting coordinates. I replaced it with a simple check for zero vectors (if both input points are the same).

File: fuzz/fuzz_targets/fuzz.rs (hunk 5 above)
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs:247]: Commented!

File: fuzz/fuzz_targets/fuzz.rs (hunk 6 above)
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs]: Done and done

File: fuzz/fuzz_targets/fuzz.rs (hunk 8 above)
Comment (finnbear) [reply fuzz/fuzz_targets/fuzz.rs:338]: This was the best thing I could think of. If this was a normal unit test, I'd give each shape a unique numerical ID and compare sets of ID's. However, an *arbitrary* shape cannot really be made to have a unique ID, so we use the pointer instead.

This method leads to good debug output when things go wrong since the shapes are actually printed.

File: fuzz/fuzz_targets/fuzz.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (svenstaro) [reply fuzz/fuzz_targets/fuzz.rs]: Excellent :+1:
